### PR TITLE
SparseMatrixCSC checks for negative dimensions and fixes for maximum and minimum

### LINKED ISSUE #10407: Error when computing the maximum of an empty sparse matrix
When trying to compute the maximum of an empty sparse matrix you get an error instead of `0.0` as a result:

In Julia version 0.3.6

    julia> maximum(spzeros(5, 5))
    ERROR: Reducing over an empty array is not allowed.
     in _mapreduce at reduce.jl:160
     in maximum at sparse/sparsematrix.jl:718

Or version 0.4.0-dev+3677

    julia> maximum(spzeros(5, 5))
    ERROR: ArgumentError: reducing over an empty collection is not allowed
     in _mapreduce at reduce.jl:135
     in maximum at sparse/sparsematrix.jl:744

The dense counterpart behaves correctly and reports `0.0` as the result

    julia> maximum(zeros(5, 5))
    0.0

### LINKED ISSUE #10411: Some sparse matrix constructors allow negative matrix dimensions
Some others give confusing errors.

```
julia> A=spzeros(-2,2) #???
-2x2 sparse matrix with 0 Float64 entries:

julia> sprand(-2,2,0.2) #Confusing error
ERROR: InexactError()
 in sprand at sparse/sparsematrix.jl:367
 in sprand at sparse/sparsematrix.jl:398

julia> speye(-2,2) # OK
ERROR: invalid Array dimensions
 in call at ./base.jl:275
 in speye at sparse/sparsematrix.jl:422
 in speye at sparse/sparsematrix.jl:414
```

### PR DESCRIPTION
**Pull request by tanmaykm**
- Refactored sparse matrix constructor and added check for dimensions. Also added dimensions check to `spzeros` and `sprand` (fixes #10411)
- Added `nnz` check to `maximum` and `minimum` (fixes #10407)

### CODE REVIEW

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -9,10 +9,17 @@ type SparseMatrixCSC{Tv,Ti<:Integer} <: AbstractSparseMatrix{Tv,Ti}
     colptr::Vector{Ti}      # Column i is in colptr[i]:(colptr[i+1]-1)
     rowval::Vector{Ti}      # Row values of nonzeros
     nzval::Vector{Tv}       # Nonzero values
-end
 
-SparseMatrixCSC{Tv,Ti}(m::Integer, n::Integer, colptr::Vector{Ti}, rowval::Vector{Ti}, nzval::Vector{Tv}) =
-    SparseMatrixCSC(int(m), int(n), colptr, rowval, nzval)
+    function SparseMatrixCSC(m::Integer, n::Integer, colptr::Vector{Ti}, rowval::Vector{Ti}, nzval::Vector{Tv})
+        ((m < 0) || (n < 0)) && error("invalid Array dimensions")
```
Comment (ViralBShah) [inline base/sparse/sparsematrix.jl]: I think we should be throwing an appropriate exception here. Something like `DimensionMismatch("Negative dimensions")`. @jakebolewski ?

File: base/sparse/sparsematrix.jl [hunk 2]
```diff
@@ -358,6 +365,7 @@ end
 
 function sprand{T}(m::Integer, n::Integer, density::FloatingPoint,
                    rng::Function,::Type{T}=eltype(rng(1)))
+    ((m < 0) || (n < 0)) && error("invalid Array dimensions")
```
Comment (ViralBShah) [inline base/sparse/sparsematrix.jl]: Throw an exception here too.

File: base/sparse/sparsematrix.jl (hunk 1 above)
Comment (jakebolewski) [inline base/sparse/sparsematrix.jl]: I would write something like

`m < 0 && throw(ArgumentError("number of rows (m) must be ≥ 0, got $m"))`
`n < 0 && throw(ArgumentError("number of columns (n) must  be ≥ 0, got $n"))`

Comment (jakebolewski) [inline base/sparse/sparsematrix.jl]: `DimensionMismatch` is used a bit inconsistently in Base, I don't know if that would be more appropriate.

Comment (ViralBShah) [inline base/sparse/sparsematrix.jl]: `ArgumentError` seems more appropriate. Do we document all the exceptions in the manual? That would help someone pick an appropriate one to throw.

**Comment by tanmaykm** [comment]
Fixed and rebased. Also added `real` and `imag` to fix #10412

**Comment by tkelman** [comment]
Thanks for splitting out `real` and `imag`, better as 2 separate PR's. The checks here lgtm.

Is there any performance impact from removing the `{Tv,Ti}` parameters from the constructor call sites?

**Comment by tanmaykm** [comment]
I'm not sure. Should we add a type assert at those places?

**Comment by ViralBShah** [comment]
Can you run the tests in test/perf/sparse and see if it makes a difference. It is possible that those annotations are not necessary any more.

**Comment by tanmaykm** [comment]
I think the changing the constructor will not affect the performance. Because the new external constructor is just a wrapper over the internal constructor, and the calling methods themselves are parameterized with `{Tv,Ti}`.

But will run the perf tests anyway.

**Comment by tanmaykm** [comment]
Didn't find any significant difference. Here are the perf test results: https://gist.github.com/tanmaykm/f98254e4343c85e3a731

**Comment by tkelman** [comment]
looks okay then. needs rebasing (`int` is now deprecated), and should `speye` and `sprand` also be throwing `ArgumentError`?

**Comment by tanmaykm** [comment]
Rebased, `speye` and `sprand` now throw `ArgumentError`.

**Comment by tkelman** [comment]
lgtm

**Comment by tkelman** [comment]
backported in 4d1b751ddaa92f6402e22a8ebeb3c9c086a1d14a

**Pull request merged**